src/util/child_common.c

```diff
@@ -542,15 +545,27 @@ static void _read_pipe_handler(struct tevent_context *ev,
             size = sss_atomic_read_s(state->fd, buf, len);
         }
     } else {
-        size = sss_atomic_read_s(state->fd, buf, CHILD_MSG_CHUNK);
+        if (state->non_blocking) {
+            size = sss_atomic_read(state->fd, buf, CHILD_MSG_CHUNK);
+        } else {
+            size = sss_atomic_read_s(state->fd, buf, CHILD_MSG_CHUNK);
+        }
     }
     if (size == -1) {
         err = errno;
+        if (state->non_blocking && (err == EWOULDBLOCK || err == EAGAIN)) {
+            DEBUG(SSSDBG_TRACE_ALL,
+                  "Waiting for more data to read, returning the event loop. Current size [%zu]\n", state->len);
+            return;
+        }
+
         DEBUG(SSSDBG_CRIT_FAILURE,
               "read failed [%d][%s].\n", err, strerror(err));
+
         tevent_req_error(req, err);
         return;
     } else if (size > 0) {
+        DEBUG(SSSDBG_TRACE_ALL, "Adding [%zd] bytes of data.\n", size);
```

Comment:
I have a feeling that original code is buggy here.

In case of `state->safe == true` `len` is initialized by reading header.
But in case of `state->safe == false` it is zero initialized and that's all, so condition `(state->len == len)` below can't be met.

So in case `state->safe == false` operation can only be completed if fatal error happens or fd is closed.
This makes sense, of course, but code
```
        if (state->len == len) {
            DEBUG(SSSDBG_TRACE_FUNC, "All data received\n");
            tevent_req_done(req);
        }
```
is misleading.

For other branch - `state->safe == true` - this code doesn't make much sense at all.

Reply:
I haven't touched this because I'm wondering as well about `state->len == len` since I would expect `state->len` giving the whole length of the received data and `len` in safe more just being the length of the messages, i.e. `state-len - sizeof(uint32_t)`. I would prefer if this can be handled in the separate PR.

Reply:
> I would prefer if this can be handled in the separate PR.

Ok.